pkg/pillar/cmd/zedagent/handleconfig.go

```diff
@@ -150,6 +152,127 @@ var nilUUID uuid.UUID
+	}
+	changed, configSha, err := fileutils.CompareSha(
+		types.BootstrapConfFileName, types.BootstrapShaFileName)
+	if err != nil {
+		log.Errorf("CompareSha failed for bootstrap config: %s", err)
+		// We will not record SHA for applied bootstrap config
+		// and as a result load it again with the next boot.
+		// However, with config timestamp preventing accidental revert
+		// to an older configuration, this should not be a problem.
+		configSha = nil
+	} else if !changed {
+		// This bootstrap config was already applied.
+		return
+	}
+
+	// Load file content.
+	contents, err := ioutil.ReadFile(types.BootstrapConfFileName)
+	if err != nil {
+		log.Errorf("Failed to read bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Mark bootstrap config as processed by storing SHA hash of its content
+	// under /persist/ingested/ directory.
+	// We do this even even if the unmarshalling (or anything else) below fails.
+	// This is to avoid repeated failing attempts to load invalid config on each boot.
+	defer func() {
+		if configSha != nil {
+			err := fileutils.SaveShaInFile(types.BootstrapShaFileName, configSha)
+			if err != nil {
+				log.Errorf("Failed to save SHA of bootstrap config: %v", err)
+			}
+		}
+	}()
+
+	// Unmarshal BootstrapConfig.
+	bootstrap := zconfig.BootstrapConfig{}
+	err = proto.Unmarshal(contents, &bootstrap)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Verify controller certificate chain.
+	sigCertBytes, err := zedcloud.VerifySigningCertChain(log, bootstrap.ControllerCerts)
+	if err != nil {
+		log.Errorf("Controller cert chain verification failed for bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Verify payload signature
+	zedcloudCtx := zedcloud.NewContext(log, zedcloud.ContextOptions{})
+	if err = zedcloud.LoadServerSigningCert(&zedcloudCtx, sigCertBytes); err != nil {
+		log.Errorf("Failed to load signing server cert from bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+	_, err = zedcloud.VerifyAuthContainer(&zedcloudCtx, bootstrap.SignedConfig)
+	if err != nil {
+		log.Errorf("Signature verification failed for bootstrap config: %v", err)
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+
+	// Unmarshal EdgeDevConfig from the payload of AuthContainer.
+	devConfig := zconfig.EdgeDevConfig{}
+	payload := bootstrap.SignedConfig.GetProtectedPayload().GetPayload()
+	if payload == nil {
+		log.Error("Bootstrap config payload is nil")
+		indicateInvalidBootstrapConfig(getconfigCtx)
+		return
+	}
+	err = proto.Unmarshal(payload, &devConfig)
+	if err != nil {
+		log.Errorf("Failed to unmarshal bootstrap config payload: %v", err)
+		return
+	}
+
+	// Apply bootstrap config.
+	inhaleDeviceConfig(getconfigCtx, &devConfig, fromBootstrap)
+	log.Notice("Bootstrap config was applied")
```

Comment:
This log entry will appear even if the bootstrap config is older than what was received from the controller.
Can we get some return value from inhale to indicate it was ignored and log things appropriately?

Reply:
Added `configProcessingRetval` enum, now used by `getLatestConfig`, `inhaleDeviceConfig` and `parseConfig`.
Logging was corrected here.